// server/api/food/food.controller.js
/**
 * GET     /api/foods              ->  index
 * POST    /api/foods              ->  create
 * GET     /api/foods/:id          ->  show
 * PUT     /api/foods/:id          ->  update
 * DELETE  /api/foods/:id          ->  destroy
 */

import assign from 'lodash/assign';
import {handleError} from 'server/api/common';
import config from '../../config/environment';
var request = require('request-promise').defaults({
  baseUrl: config.ndb.baseUrl,
});
var qs = {api_key: config.ndb.apiKey};

function respondWithResult(res, statusCode) {
  statusCode = statusCode || 200;
  return function onSuccess(entity) {
    if(entity) {
      res.status(statusCode).json(JSON.parse(entity));
    }
  };
}

// Gets a list of Foods
export function index(req, res) {
  request.get({
    url: '/list',
    qs: assign(qs, {
      lt: 'f',
      max: req.query.max || 50,
      offset: req.query.offset || 0,
    }),
  })
    .then(respondWithResult(res))
    .catch(handleError(res));
}

export function show(req, res) {
  request.get({
    url: '/reports',
    qs: assign(qs, {
      ndbno: req.params.id,
    }),
  })
    .then(respondWithResult(res))
    .catch(handleError(res));
}
