// client/app/components/foodlog/AddMeal.jsx
import React from 'react';
import {Form, FormGroup, InputGroup, FormControl, Button, Glyphicon} from 'react-bootstrap';
import IngredientsAutocomplete from 'client/app/components/shared/IngredientsAutocomplete';

import './AddMeal.scss';

const DEFAULT_QUANTITY=100;
const INITIAL_STATE = {selectedIngredient: {name: '', ndbno: '-1'}, quantity: ''};

export default class AddMealContainer extends React.Component {
  constructor(...args) {
    super(...args);

    this.state = INITIAL_STATE;

    this.ingredientSelected = this.ingredientSelected.bind(this);
    this.buttonClicked = this.buttonClicked.bind(this);
    this.quantityChanged = this.quantityChanged.bind(this);
  }

  ingredientSelected(value) {
    if(value) {
      this.setState({selectedIngredient: {name: value.name, ndbno: value.ndbno}});
    } else {
      this.setState(INITIAL_STATE);
    }
  }

  buttonClicked() {
    if(this.state.selectedIngredient.name) {
      var quantity = this.state.quantity ? parseFloat(this.state.quantity) : DEFAULT_QUANTITY;
      if(isNaN(quantity)) {
        quantity = DEFAULT_QUANTITY;
      }

      this.props.onAdd(this.state.selectedIngredient, quantity);
    }
    this.setState(INITIAL_STATE);
    this.refs.autocomplete.clearSelection();
  }

  quantityChanged(e) {
    this.setState({quantity: (e.target.value ? e.target.value : '')});
  }

  render() {
    return (
      <Form inline>
        <FormGroup className="s__AddMeal">
          <IngredientsAutocomplete ref="autocomplete" onIngredientSelected={this.ingredientSelected}/>

          <InputGroup className="s__foodlog_quantity" >
            <FormControl value={this.state.quantity}
                         placeholder={DEFAULT_QUANTITY.toString()}
                         onChange={this.quantityChanged}/>
            <InputGroup.Addon>g</InputGroup.Addon>
          </InputGroup>

          <Button onClick={this.buttonClicked} aria-label="Add">
            <Glyphicon glyph="plus" />
          </Button>
        </FormGroup>
      </Form>
    );
  }
}

AddMealContainer.propTypes = {
  placeholder: React.PropTypes.string,
  onAdd: React.PropTypes.func,
};

AddMealContainer.defaultProps = {
  placeholder: '',
  onAdd: (ingredient, quantity) => {}
};
