// server/config/environment/index.js
'use strict';

import path from 'path';
import _ from 'lodash';

function ensureEnv(name) {
  if(!process.env[name]) {
    throw new Error(`You must set the ${name} environment variable`);
  }
}

ensureEnv('NODE_ENV');
ensureEnv('SERVER_ROOT');

// All configurations will extend these options
const all = {
  env: process.env.NODE_ENV,

  // Root path of server
  root: path.normalize(process.env.SERVER_ROOT),

  // Root path of the client
  clientRoot: process.env.CLIENT_ROOT ? path.normalize(process.env.CLIENT_ROOT) : null,

  // Server port
  port: process.env.PORT || 9000,

  // Server IP
  ip: process.env.IP || '0.0.0.0',

  // Should we populate the DB with sample data?
  seedDB: false,

  // Secret for session, you will want to change this and make it an environment variable
  secrets: {
    session: 'food-diary-secret',
  },

};

// Export the config object based on the NODE_ENV
module.exports = _.merge(
  all,
  require('./shared'),
  require(`./${process.env.NODE_ENV}.js`) || {});
