// client/components/App.jsx

import React from 'react';

import NotesView from './NotesView';
import LoginView from './LoginView';

export default class App extends React.Component {
  constructor({cradle}) {
    super();
    this.cradle = cradle;
    this.actionHolder = this.cradle.appActions;
    this.store = this.cradle.appStore;

    this.state = {};
    this.state.storeState = this.store.getState();
    this.onStoreChanged = this.onStoreChanged.bind(this)
  }

  componentDidMount() {
    this.store.listen(this.onStoreChanged);
  }

  componentWillUnmount() {
    this.store.unlisten(this.onStoreChanged);
  }

  onStoreChanged(state) {
    this.setState({storeState: state});
  }

  render() {
    if (this.state.storeState.loggedIn) {
      return (
        <NotesView
          actions={this.cradle.notesActions}
          store={this.cradle.notesStore}
        />);
    } else {
      return (
        <LoginView
          onLogin={(username, password) => this.actionHolder.login({username, password})}
        />)
    }
  }
};
