// client/components/HelloWorld.jsx
import React from 'react';

import './HelloWorld.scss';
import Button from 'react-bootstrap/lib/Button';
import Jumbotron from 'react-bootstrap/lib/Jumbotron';

function FetchedExample(props) {
  if(props.data.example) {
    return (<p> {props.data.example} </p>);
  } else {
    return (<p> Error: {props.data.error} </p>);
  }
}

export default class HelloWorld extends React.Component {
  constructor(props) {
    super();
    this.actionHolder = props.actions;
    this.store = props.store;

    this.state = props.store.getState();
    this.onStoreChanged = this.onStoreChanged.bind(this)
  }

  componentDidMount() {
    this.store.listen(this.onStoreChanged);

    this.actionHolder.fetchExample();
  }

  componentWillUnmount() {
    this.store.unlisten(this.onStoreChanged);
  }

  onStoreChanged(state) {
    this.setState(state);
  }

  render() {
    return (
      <Jumbotron>
        <h1>Hello, world!</h1>
        <p>
          Simple jumbotron-style component for calling extra attention to featured content.
        </p>
        <p><Button bsStyle="primary">Enroll</Button></p>
        <FetchedExample data={this.state}/>
      </Jumbotron>
    );
  }
};
