// client/components/NotesView.jsx
import React from 'react';

import ListGroup from 'react-bootstrap/lib/ListGroup';
import ListGroupItem from 'react-bootstrap/lib/ListGroupItem';

function NotesList(props) {
  const items = [];

  for(let itemId = 0; itemId < props.list.length; itemId += 1) {
    items.push(
      <ListGroupItem key={itemId} onClick={() => props.onSelect(itemId)}>
        {props.list[itemId].title}
      </ListGroupItem>
    )
  }

  return (
    <ListGroup>
      {items}
    </ListGroup>
  );
}

function NoteEditor(props) {
 return (<p>Editor</p>);
}

export default class NotesView extends React.Component {
  constructor(props) {
    super();
    this.actionHolder = props.actions;
    this.store = props.store;

    this.state = {};
    this.state.storeState = props.store.getState();
    this.onStoreChanged = this.onStoreChanged.bind(this)
  }

  componentDidMount() {
    this.store.listen(this.onStoreChanged);

    this.actionHolder.fetch();
  }

  componentWillUnmount() {
    this.store.unlisten(this.onStoreChanged);
  }

  onStoreChanged(state) {
    this.setState({storeState: state});
  }

  render() {
    const notesList = (<NotesList
      list={this.state.storeState.notesList}
      onSelect={(id) => this.actionHolder.select(id)}
      onDeselect={() => this.actionHolder.deselect()}
    />);

    if(this.state.storeState.selectedNote) {
      return (
        <div>
          {notesList}
          <NoteEditor />
        </div>
      );
    } else {
      return notesList;
    }
  }
};
