// client/app/components/navbar/Navbar.jsx
import _ from 'lodash';

import React from 'react';
import {LinkContainer} from 'react-router-bootstrap';
import {Navbar, Nav, NavItem} from 'react-bootstrap';

import Omnibox from './Omnibox';
import NotificationDropown from './NotificationDropdown';
import SettingsDropdown from './SettingsDropdown';
import IconNavItem from './IconNavItem';

import {logout} from 'client/sharedjs/auth';

import './navbar.scss';

export default class MainNavbar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false,
    };
    _.bindAll(this, ['collapseNav', 'toggleExpanded', 'onSearch']);
  }

  collapseNav() {
    this.setState({expanded: false});
  }

  onSearch(searchTerm) {
    this.collapseNav();
    this.props.onSearch(searchTerm);
  }

  toggleExpanded() {
    this.setState({expanded: !this.state.expanded});
  }

  render() {
    return (
      <Navbar expanded={this.state.expanded} onToggle={this.toggleExpanded}>
        <Navbar.Header>
          <Navbar.Brand>
            {/* eslint-disable jsx-a11y/href-no-hash */}
            <a href="#">BiteTracker</a>
            {/* eslint-enable jsx-a11y/href-no-hash */}
          </Navbar.Brand>

          <Navbar.Toggle />
        </Navbar.Header>

        <Navbar.Collapse>

          <Navbar.Form pullLeft>
            <Omnibox onSearch={this.onSearch} />
          </Navbar.Form>

          {this.props.loggedIn ? (
            <Nav pullRight>
              {/* Standard navigation links */}
              <LinkContainer to="/recipes">
                <NavItem eventKey={1} onClick={this.collapseNav}>Recipes</NavItem>
              </LinkContainer>
              <LinkContainer to="/log">
                <NavItem eventKey={2} onClick={this.collapseNav}>Food Log</NavItem>
              </LinkContainer>
              <LinkContainer to="/planner">
                <NavItem eventKey={3} onClick={this.collapseNav}>Meal Planner</NavItem>
              </LinkContainer>
              <LinkContainer to="/groceries">
                <NavItem eventKey={4} onClick={this.collapseNav}>Groceries</NavItem>
              </LinkContainer>

              {/* Notification dropdown - only visible on larger screeens */}
              <NotificationDropown eventKey={5} />

              {/* Icon menu items */}
              <LinkContainer to="/notifications">
                <IconNavItem glyph="bell"
                             text="Notifications"
                             className="navbar-notif-link"
                             eventKey={5}
                             onClick={this.collapseNav} />
              </LinkContainer>

              <LinkContainer to="/profile">
                <IconNavItem glyph="user"
                             text="Profile"
                             eventKey={6}
                             onClick={this.collapseNav} />
              </LinkContainer>

              <LinkContainer to="/settings">
                <IconNavItem glyph="cog"
                             text="Settings"
                             eventKey={7}
                             onClick={this.collapseNav} />
              </LinkContainer>

              <IconNavItem glyph="log-out"
                           text="Logout"
                           onClick={logout}
                           eventKey={8} />

              {/* Settings dropdown - only visible on small devices, replaces menu icons above */}
              <SettingsDropdown eventKey={9} />
            </Nav>
          ) : null}

        </Navbar.Collapse>
      </Navbar>
    );
  }
}

MainNavbar.propTypes = {
  loggedIn: React.PropTypes.bool.isRequired,
  onSearch: React.PropTypes.func.isRequired
};
