// client/app/components/shared/SearchBox.jsx
import React from 'react'

import { FormGroup, FormControl, InputGroup, Button, Glyphicon } from 'react-bootstrap'

export default class SearchBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = { searchTerm: '' };
    this.onChange = this.onChange.bind(this);
    this.onSearch = this.onSearch.bind(this);
  }

  onChange(e) {
    this.setState({ searchTerm: e.target.value });
  }

  onSearch(e) {
    e.preventDefault();
    this.props.onSearch(this.state.searchTerm);
  }

  render() {
    return (
      <form>
        <FormGroup>
          <InputGroup>
            <FormControl type="text" placeholder={this.props.placeholder} onChange={this.onChange} />
            <InputGroup.Button>
              <Button type="submit" onClick={this.onSearch}>
                <Glyphicon glyph="search" aria-label={this.props.placeholder} />
              </Button>
            </InputGroup.Button>
          </InputGroup>
        </FormGroup>
      </form>
    );
  }
}

SearchBox.propTypes = {
  placeholder: React.PropTypes.string,
  onSearch: React.PropTypes.func.isRequired
};
