// server/config/express.js
/**
 * Express configuration
 */

'use strict';

import express from 'express';
import morgan from 'morgan';
import bodyParser from 'body-parser';
import methodOverride from 'method-override';
import errorHandler from 'errorhandler';
import lusca from 'lusca';
import config from './environment';


export default function(app) {
  const env = app.get('env');

  app.use(bodyParser.urlencoded({extended: false}));
  app.use(bodyParser.json());
  app.use(methodOverride());

  /**
   * Lusca - express server security
   */
  if(env !== 'test') {
    app.use(lusca({
      csrf: false,
      xframe: 'SAMEORIGIN',
      hsts: {
        maxAge: 31536000, // 1 year, in seconds
        includeSubDomains: true,
        preload: true,
      },
      xssProtection: true,
    }));
  }

  if(env === 'development') {
    app.use(require('connect-livereload')());
  }

  app.use(morgan('dev'));

  if(env === 'development' || env === 'test') {
    app.use(errorHandler()); // Error handler - has to be last
  }
}
