// server/config/index.js
'use strict';

import path from 'path';
import _ from 'lodash';
import fs from 'fs';

function ensureEnv(name) {
  if(!process.env[name]) {
    throw new Error(`You must set the ${name} environment variable`);
  }
}

ensureEnv('NODE_ENV');
ensureEnv('SERVER_ROOT');

// All configurations will extend these options
const frontendConfigPath = "/etc/appconfig/frontend-config.json";
var frontendConfig = {};

try {
  frontendConfig = JSON.parse(fs.readFileSync(frontendConfigPath, "utf8"));
  console.log(frontendConfig);
} catch(e) {
  console.log(`Cannot file frontend configuration file at ${frontendConfigPath}\nException: ${e}`);
}

const clientRoot = process.env.CLIENT_ROOT ? path.normalize(process.env.CLIENT_ROOT) : null;

const all = {
  env: process.env.NODE_ENV,

  // Root path of server
  root: path.normalize(process.env.SERVER_ROOT),

  // Root path of the client
  clientRoot,

  htmlTemplate: clientRoot + "/index.html",

  // Server port
  port: process.env.PORT || 9000,

  // Server IP
  ip: process.env.IP || '0.0.0.0',

  frontendConfig,
};

// Export the config object based on the NODE_ENV
module.exports = _.merge(
  all,
  require(`./environment-specific/${process.env.NODE_ENV}.js`) || {});
