// client/app/components/shared/form/Checkbox.jsx
import uniqueId from 'lodash/uniqueId';
import classNames from 'classnames';
import React, {PropTypes} from 'react';

import './checkbox.scss';

const ID_PREFIX = 'checkbox-';

export default class Checkbox extends React.Component {
  componentWillMount() {
    this.id = this.props.id || uniqueId(ID_PREFIX);
  }

  render() {
    const {id: _, bsStyle, bsSize, rounded, validationState, children, ...inputProps} = this.props;
    return (
      <div className={classNames({
        's__checkbox': true,
        [`checkbox-${bsStyle}`]: true,
        [`checkbox-${bsSize}`]: bsSize,
        'checkbox-circle': rounded,
        [validationState]: validationState,
      })}>
        <label htmlFor={this.id}>
          <input type="checkbox"
                 id={this.id}
                 {...inputProps} />
          {children}
        </label>
      </div>
    );
  }
}

Checkbox.propTypes = {
  bsStyle: PropTypes.string,
  bsSize: PropTypes.string,
  rounded: PropTypes.bool,
  id: PropTypes.string,
  validationState: PropTypes.string,
  children: PropTypes.node,
};

Checkbox.defaultProps = {
  bsStyle: 'primary',
};
